lib/server.js

```diff
@@ -0,0 +1,205 @@
+    } catch (err) {
+      return this.write(res, 403, 'Bad signature')
+    }
+
+    try {
+      var validUrl = await SafeHttpClient.checkUrl(decodedUrl)
+    } catch (err) {
+      // @ts-ignore err does have message
+      return this.write(res, 400, err.message)
+    }
+
+    const controller = new AbortController()
+    const {signal} = controller
+
+    req.on('close', () => {
+      controller.abort()
+    })
+
+    try {
+      // TODO: respect forwarded headers (check if not private IP)
+      const filterRequestHeaders = filterHeaders(REQUEST_HEADERS)
+      const filterResponseHeaders = filterHeaders(RESPONSE_HEADERS)
+      const client = new SafeHttpClient(this.config.maxSize)
+      const {buffer, headers: resHeaders} = await client.safeFetch(validUrl, {
+        // @ts-ignore we can convert req.headers to Headers
+        headers: filterRequestHeaders(new Headers(req.headers)),
+        method: req.method,
+        // We can't blindly follow redirects as the initial checkUrl
+        // might have been safe, but the redirect location might not be.
+        // SafeHttpClient will check the redirect location before following it.
+        redirect: 'manual',
+        signal
+      })
+
+      const headers = {
+        ...DEFAULT_HEADERS,
+        ...filterResponseHeaders(resHeaders),
+        Via: this.config.serverName
+      }
+
+      if (req.method === 'HEAD') {
+        res.writeHead(204, headers)
+        return res.end()
+      }
+      res.writeHead(200, headers)
+      res.write(buffer)
+      return res.end()
+    } catch (err) {
+      if (err instanceof HttpError) {
+        return this.write(res, err.statusCode, err.message)
+      }
+      if (err instanceof Error) {
+        if (err.name === 'AbortError') {
+          return
+        }
+        const msg = err.message || 'Internal server error'
+        return this.write(res, 500, msg)
+      }
+    }
+  }
+
+  /**
+   * @param {string} receivedDigest
+   * @param {string} hex
+   * @private
+   */
+  verifyHMAC(receivedDigest, hex) {
+    // Hex-decode the URL
+    const originalUrlBuffer = hexDecode(hex)
+
+    // Convert the buffer to a string
+    const decodedUrl = originalUrlBuffer.toString('utf-8')
+
+    // Verify the HMAC digest to ensure the URL hasn't been tampered with
+    const hmac = crypto.createHmac('sha1', this.config.HMACKey)
+    hmac.update(decodedUrl)
+    const expectedDigest = hmac.digest('hex')
+
+    if (expectedDigest !== receivedDigest) {
+      throw new Error('URL integrity check failed')
+    }
+
+    return decodedUrl
+  }
+
+  /**
+   * @param {http.ServerResponse} res
+   * @param {number} status
+   * @param {unknown} [body]
+   * @param {Record<string, string | undefined>} [headers]
+   * @private
+   */
+  write(res, status, body, headers = DEFAULT_HEADERS) {
+    if (status !== 204) {
+      // @ts-expect-error not explicitly typed but possible
+      headers['Content-Length'] = Buffer.byteLength(body, 'utf8')
+    }
```

Comment:
```suggestion
   * @param {Buffer | string} body
   * @param {Record<string, string | undefined>} [headers]
   * @private
   */
  write(res, status, body, headers = DEFAULT_HEADERS) {
    if (status !== 204) {
      headers['Content-Length'] = String(Buffer.byteLength(body, 'utf8'))
    }
```

Not sure what the actual valid `body`s are that you expect, but `Buffer | string` probably allows most cases? Is it allowed to be `undefined`? These are the ones allowed by `Buffer.byteLength`: `NodeJS.ArrayBufferView | ArrayBuffer | SharedArrayBuffer | string`.